### PR TITLE
add uuid in cudaDeviceProperties

### LINKED ISSUE #99903: Can the CUDA device LUID be exposed as part of _CudaDeviceProperties?
### 🚀 The feature, motivation and pitch

CUDA devices can be ordered in a few different ways (FASTEST_FIRST, PCI_BUS_ID). The device LUID is the more modern way of describing a GPU in a unique way. It's particularly useful when using CUDA along with other APIs such as Vulkan or DirectX.
Can the LUID that the cudaDeviceProp structure contains be made visible in _CudaDeviceProperties as well?
https://docs.nvidia.com/cuda/cuda-runtime-api/structcudaDeviceProp.html#structcudaDeviceProp_1b1fe931f3f41d1b97c8679c8b15d94e3

Thanks

### Alternatives

_No response_

### Additional context

_No response_

cc @ngimel

### PR DESCRIPTION
**Pull request by jeffdaily**
Replaces #99967.

Fixes #99903.

cc @ptrblck @msaroufim @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

**Comment by cpuhrsch** [comment]
There's a couple failing tests. Hud indicates errors are related https://hud.pytorch.org/pr/125083

**Comment by jeffdaily** [comment]
> There's a couple failing tests. Hud indicates errors are related https://hud.pytorch.org/pr/125083

Apologies, I should have moved it back to draft after my first commit once I realized this wasn't as straightforward as anticipated.  First commit passed, but there was no unit test coverage.  Adding a unit test proved the first try was wrong.  Many commits later, finally got the correct implementation and linting.  Please review now.

File: torch/csrc/cuda/Module.cpp
Comment (albanD) [inline torch/csrc/cuda/Module.cpp:916]: Can we read the size of cudaUUID_t here to reduce hard-coding?

Comment (jeffdaily) [reply torch/csrc/cuda/Module.cpp:916]: UUIDs are a 128-bit label.  CUDA and HIP store this as char[16].  For string representation, the code here expands this to 8-4-4-4-12 hex format, so it will be 32 ascii + 4 hyphens + 1 for the null byte.

If it helps I could add the above as a comment to indicate why the size is hard-coded?

Comment (albanD) [reply torch/csrc/cuda/Module.cpp:916]: I guess that would be sizeof(cudaUUID_t) * 2 + 4 + 1 then? And with the comment to explain the 4 and 1. Sounds good!

File: torch/csrc/cuda/Module.cpp
Comment (albanD) [inline torch/csrc/cuda/Module.cpp:912]: Where is the CUuuid class defined?

Comment (jeffdaily) [reply torch/csrc/cuda/Module.cpp:912]: cuda.h has the following:

```C
#ifndef CU_UUID_HAS_BEEN_DEFINED
#define CU_UUID_HAS_BEEN_DEFINED
typedef struct CUuuid_st {                                /**< CUDA definition of UUID */
    char bytes[16];
} CUuuid;
#endif
```

and cuda's driver_types.h has this:

```C
/**
 * CUDA UUID types
 */
#ifndef CU_UUID_HAS_BEEN_DEFINED
#define CU_UUID_HAS_BEEN_DEFINED
struct __device_builtin__ CUuuid_st {     /**< CUDA definition of UUID */
    char bytes[16];
};
typedef __device_builtin__ struct CUuuid_st CUuuid;
#endif
typedef __device_builtin__ struct CUuuid_st cudaUUID_t;
```

Comment (albanD) [reply torch/csrc/cuda/Module.cpp:912]: Ho right, so the whole thing is typedefed to cudaUUID_t that is the one documented at https://docs.nvidia.com/cuda/cuda-runtime-api/structcudaDeviceProp.html#structcudaDeviceProp_1626c20637498c7be1381db55a6261308
Should we use cudaUUID_t everywhere here?

File: torch/csrc/cuda/Module.cpp
Comment (jeffdaily) [reply torch/csrc/cuda/Module.cpp]: Done.

**Comment by stellaraccident** [comment]
Thanks, Jeff. Nit: Maybe a class level comment which says that the str format is expected to match the format used by nvidia-smi and can be counted on for that (i.e. it is stable and will only be updated if future updates require it to match a different format from that tool)? If doing that, I think it would be more appropriate to implement as `__str__` vs `__repr__` (since repr is usually just for debugging).

**Comment by jeffdaily** [comment]
> Thanks, Jeff. Nit: Maybe a class level comment which says that the str format is expected to match the format used by nvidia-smi and can be counted on for that (i.e. it is stable and will only be updated if future updates require it to match a different format from that tool)? If doing that, I think it would be more appropriate to implement as `__str__` vs `__repr__` (since repr is usually just for debugging).

I opted to drop the "GPU-" prefix that would normally appear with nvidia-smi; the way I have it implemented here is a pure UUID format.  Would it be better to have left the GPU- prefix as part of the string?  Since this PR hasn't had maintainer approval yet, we have time to put the GPU- prefix back in if you think it is more appropriate.  And I can certainly swap __str__ for __repr__ too.

**Comment by albanD** [comment]
I would be curious if @ptrblck have a strong opinion about keeping the GPU- header here? Or any other comment?

**Review by pruthvistony — APPROVED** [review]

**Review by albanD — APPROVED** [review]
Ok SGTM
We can revisit if there is any new opinion from @eqy later.

**Review by eqy — APPROVED** [review]

**Comment by jeffdaily** [comment]
> @pytorchbot revert -m "test_uuid is flaky? ex https://github.com/pytorch/pytorch/actions/runs/8988855916/job/24692369523 https://hud.pytorch.org/flakytest?name=test_uuid&suite=TestCuda&file=%25&limit=300" -c nosignal

So sorry, I thought all was well.  I will take another crack at this.  Thanks for the link to the failed jobs.

File: test/test_cuda.py
Comment (albanD) [inline test/test_cuda.py:280]: Ho completely missed that... We should use self.assertEqual here. That would have given us much nice error message to work with from the test job!

Comment (jeffdaily) [reply test/test_cuda.py]: Thanks for the suggestion.  I have changed it to self.asserEqual now.

**Comment by jeffdaily** [comment]
Instead of uuid returning bytes, I switched it to explicitly return a std::vector<int8_t> which is interpreted as a list of ints.  It will always be len 16.

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by izaitsevfb** [comment]
@pytorchbot revert -m "Fails internal builds with: no member named 'uuid' in 'hipDeviceProp_t'" -c ghfirst

Looks like the same [issue as here](https://github.com/pytorch/pytorch/pull/99967#issuecomment-1624238695), how was it fixed?

```
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:989:49: error: no member named 'uuid' in 'hipDeviceProp_t'
               << ", uuid=" << std::string(prop.uuid.bytes, 16) << ")";
                                           ~~~~ ^
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:979:47: error: no member named 'uuid' in 'hipDeviceProp_t'
      .def_readonly("uuid", &hipDeviceProp_t::uuid)
                             ~~~~~~~~~~~~~~~~~^
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:982:9: error: variable 'stream' cannot be implicitly captured in a lambda with no capture-default specified
        stream << "_CudaDeviceProperties(name='" << prop.name
        ^
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:981:28: note: 'stream' declared here
        std::ostringstream stream;
                           ^
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:980:24: note: lambda expression begins here
      .def("__repr__", [](const hipDeviceProp_t& prop) {
                       ^
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:980:25: note: capture 'stream' by reference
      .def("__repr__", [](const hipDeviceProp_t& prop) {
                        ^
                        &stream
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:980:25: note: default capture by reference
      .def("__repr__", [](const hipDeviceProp_t& prop) {
                        ^
                        &
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:990:16: error: variable 'stream' cannot be implicitly captured in a lambda with no capture-default specified
        return stream.str();
               ^
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:981:28: note: 'stream' declared here
        std::ostringstream stream;
                           ^
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:980:24: note: lambda expression begins here
      .def("__repr__", [](const hipDeviceProp_t& prop) {
                       ^
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:980:25: note: capture 'stream' by reference
      .def("__repr__", [](const hipDeviceProp_t& prop) {
                        ^
                        &stream
buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/__fb_C_impl_hipify_gen_eqsb_torch/csrc/cuda/Module.cpp__/out/torch/csrc/cuda/Module.cpp:980:25: note: default capture by reference
      .def("__repr__", [](const hipDeviceProp_t& prop) {
                        ^
                        &
4 errors generated.
```

For Meta folks, see [D57138027](https://www.internalfb.com/diff/D57138027)

**Comment by jeffdaily** [comment]
Ouch, 2 reverts.  Can any Meta folks help shepherd this one?  The uuid attribute was added to hipDeviceProp_t in ROCm 6.0.  This implies Meta isn't yet up to that version of ROCm yet internally.  I could perhaps add some CMake config-time check for the attribute and some #ifdef protection and an alternative implementation.  Please advise.

**Review by malfet — APPROVED** [review]

**Comment by jeffdaily** [comment]
@malfet is there some #ifdef I could add that would remove this feature for your internal builds?  I would like to land this but I don't know how to work around your internal ROCm builds failing with this PR.

**Comment by stellaraccident** [comment]
> @malfet is there some #ifdef I could add that would remove this feature for your internal builds? I would like to land this but I don't know how to work around your internal ROCm builds failing with this PR.

+1 - we've got multiple AMD GPU integrations stalling on the inability to tie device identifiers together concretely and would really like to see this PR merged (and stay merged).

**Comment by jeffdaily** [comment]
> > @malfet is there some #ifdef I could add that would remove this feature for your internal builds?  I would like to land this but I don't know how to work around your internal ROCm builds failing with this PR.
> 
> FBCODE_CAFFE2 sounds like the one you should use, but what exactly is the problem with the internal builds?

[See earlier comment why this PR was reverted](https://github.com/pytorch/pytorch/pull/125083#issuecomment-2103315320).  It's a build issue.  The uuid attribute was added for ROCm 6.0 as a backward-breaking change as part of the major release bump.  Your internal build seems to be based on a version of ROCm that predates this change.  I don't have visibility into why your internal ROCm version isn't >= 6.0 and I don't want this PR to force you to upgrade.  That's why I was asking for a work-around.  Unless there's something else you can do as part of your phabricator import process to help land this.

**Comment by jeffdaily** [comment]
@malfet please double-check my use of FBCODE in https://github.com/pytorch/pytorch/pull/125083/commits/bd9edcaeffe0b771dc56e178f0da9d5e4b2a7154.

Any chance we could double-check if this will break Meta-internal builds before landing again?

File: torch/csrc/cuda/Module.cpp
Comment (jeffdaily) [reply torch/csrc/cuda/Module.cpp:913]: I wish I could but your internal build is at a ROCm version that claims to be 6.0 but isn't quite.  TORCH_HIP_VERSION would be >= 600 but the uuid attribute would still not be present.

**Comment by jeffdaily** [comment]
@malfet any progress landing this?

**Comment by jeffdaily** [comment]
@malfet ping, any progress landing this?

**Comment by stellaraccident** [comment]
> @malfet ping, any progress landing this?

+1 this is blocking some integrations we are doing

**Pull request closed without merging**
